File: bindings/cpp/src/client_index.cpp
#include <cstring>

#include "argv.hpp"
#include "kevy/client.hpp"
#include "kevy/errors.hpp"
#include "reply_util.hpp"

// Declarative secondary indexes: IDX.* (contract §3.8). Remote-only — the
// embedded backend answers Unsupported. Two-layered: typed shortcuts plus
// *_raw argv passthroughs that keep every server capability reachable.

namespace kevy {

using detail::i2s;
using detail::raise_reply_error;
using detail::raise_unexpected;
using detail::u2s;

namespace {

const char* idx_type_tag(IdxType t) {
  switch (t) {
    case IdxType::F64: return "f64";
    case IdxType::Str: return "str";
    default: return "i64";
  }
}

IdxPage parse_idx_page(const Reply& r) {
  if (r.kind != ReplyKind::Array || r.array.size() != 2)
    throw ProtocolError("IDX.QUERY page: expected [cursor, rows]");
  IdxPage page;
  const Reply& cur = r.array[0];
  if (cur.kind != ReplyKind::Bulk) raise_unexpected(cur);
  if (cur.bytes != "0") page.cursor = cur.bytes;
  const Reply& flat = r.array[1];
  if (flat.kind != ReplyKind::Array) throw ProtocolError("IDX.QUERY page: rows not an array");
  if (flat.array.size() % 2 != 0) throw ProtocolError("IDX.QUERY page: odd row pair");
  for (size_t i = 0; i + 1 < flat.array.size(); i += 2) {
    const Reply& k = flat.array[i];
    const Reply& v = flat.array[i + 1];
    if (k.kind != ReplyKind::Bulk || v.kind != ReplyKind::Bulk)
      throw ProtocolError("IDX.QUERY page: non-bulk row pair");
    page.rows.push_back(IdxRow{k.bytes, v.bytes});
  }
  return page;
}

std::vector<Ranked> parse_ranked(const Reply& r) {
  if (r.kind != ReplyKind::Array) raise_unexpected(r);
  std::vector<Ranked> out;
  out.reserve(r.array.size());
  for (const auto& row : r.array) {
    if (row.kind != ReplyKind::Array || row.array.size() < 2)
      throw ProtocolError("IDX.QUERY ranked row: expected [key, score]");
    if (row.array[0].kind != ReplyKind::Bulk) raise_unexpected(row.array[0]);
    out.push_back(Ranked{row.array[0].bytes, detail::score_of(row.array[1])});
  }
  return out;
}

IdxInfo parse_idx_info(const Reply& entry) {
  if (entry.kind != ReplyKind::Array) raise_unexpected(entry);
  IdxInfo info;
  const auto& cells = entry.array;
  for (size_t i = 0; i + 1 < cells.size(); i += 2) {
    const Reply& label = cells[i];
    const Reply& value = cells[i + 1];
    if (label.kind != ReplyKind::Bulk || value.kind != ReplyKind::Bulk) continue;
    const std::string& l = label.bytes;
    if (l == "name") info.name = value.bytes;
    else if (l == "prefix") info.prefix = value.bytes;
    else if (l == "kind") info.kind = value.bytes;
    else if (l == "state") info.state = value.bytes;
    else if (l == "entries") info.entries = detail::parse_uint_bytes(value.bytes);
    else if (l == "bytes") info.bytes = detail::parse_uint_bytes(value.bytes);
    // else: forward-compatible — skip unknown labels
  }
  return info;
}

}  // namespace

void Client::idx_create_range(std::string_view name, std::string_view prefix, std::string_view field,
                              IdxType ty) {
  idx_create_raw({std::string(name), "ON", "PREFIX", std::string(prefix), "FIELD", std::string(field),
                  "TYPE", idx_type_tag(ty), "KIND", "range"});
}

void Client::idx_create_raw(const std::vector<std::string>& args) {
  require_remote("IDX.CREATE");
  std::vector<std::string> argv;
  argv.reserve(args.size() + 1);
  argv.emplace_back("IDX.CREATE");
  for (const auto& a : args) argv.push_back(a);
  exec_ok(argv);
}

bool Client::idx_drop(std::string_view name) {
  require_remote("IDX.DROP");
  return exec_bool({"IDX.DROP", std::string(name)});
}

std::vector<IdxInfo> Client::idx_list() {
  require_remote("IDX.LIST");
  Reply r = exec({"IDX.LIST"});
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  if (r.kind != ReplyKind::Array) raise_unexpected(r);
  std::vector<IdxInfo> out;
  out.reserve(r.array.size());
  for (const auto& e : r.array) out.push_back(parse_idx_info(e));
  return out;
}

Reply Client::idx_query_raw(const std::vector<std::string>& args) {
  require_remote("IDX.QUERY");
  std::vector<std::string> argv;
  argv.reserve(args.size() + 1);
  argv.emplace_back("IDX.QUERY");
  for (const auto& a : args) argv.push_back(a);
  Reply r = exec(argv);
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  return r;
}

IdxPage Client::idx_query_range(std::string_view name, std::string_view min, std::string_view max,
                                uint64_t limit, std::optional<std::string_view> cursor) {
  std::vector<std::string> args{std::string(name), "RANGE", std::string(min), std::string(max),
                                "LIMIT", u2s(limit)};
  if (cursor.has_value()) {
    args.emplace_back("CURSOR");
    args.emplace_back(*cursor);
  }
  return parse_idx_page(idx_query_raw(args));
}

IdxPage Client::idx_query_eq(std::string_view name, std::string_view value, uint64_t limit) {
  return parse_idx_page(idx_query_raw({std::string(name), "EQ", std::string(value), "LIMIT", u2s(limit)}));
}

std::vector<Ranked> Client::idx_query_match(std::string_view name, std::string_view text, uint64_t limit) {
  return parse_ranked(idx_query_raw({std::string(name), "MATCH", std::string(text), "LIMIT", u2s(limit)}));
}

std::vector<Ranked> Client::idx_query_knn(std::string_view name, const std::vector<float>& vector,
                                          uint64_t k) {
  // Encode the query vector as a little-endian f32 blob (§7).
  std::string blob(vector.size() * 4, '\0');
  for (size_t i = 0; i < vector.size(); i++) {
    uint32_t bits;
    std::memcpy(&bits, &vector[i], 4);
    blob[i * 4 + 0] = static_cast<char>(bits & 0xFF);
    blob[i * 4 + 1] = static_cast<char>((bits >> 8) & 0xFF);
    blob[i * 4 + 2] = static_cast<char>((bits >> 16) & 0xFF);
    blob[i * 4 + 3] = static_cast<char>((bits >> 24) & 0xFF);
  }
  return parse_ranked(idx_query_raw({std::string(name), "KNN", blob, "LIMIT", u2s(k)}));
}

}  // namespace kevy

File: bindings/go/client_index.go
package kevy

import (
	"context"
	"encoding/binary"
	"math"
)

// Declarative secondary indexes: IDX.* (contract §3.8). Remote-only — the
// embedded backend answers Unsupported (its wire face coerces query bounds
// through the index's declared type server-side, which the client can't
// replicate without the catalog). Two-layered: typed shortcuts plus *Raw
// argv passthroughs that keep every server capability reachable.

// IdxRow is one IDX.QUERY hit: the row key plus the indexed value's wire
// string form (contract §4.3).
type IdxRow struct {
	Key   []byte
	Value []byte
}

// IdxPage is one page of IDX.QUERY RANGE/EQ results (contract §4.4).
type IdxPage struct {
	// Cursor is nil when the scan is complete (wire "0"); otherwise pass
	// it back via IdxQueryRange to resume.
	Cursor []byte
	// Rows are the hits in (value, key) order.
	Rows []IdxRow
}

// IdxInfo is one IDX.LIST entry (contract §4.5).
type IdxInfo struct {
	Name    []byte
	Prefix  []byte
	Kind    string // "range" | "unique" | "text" | "ann" | "agg"
	State   string // "ready" | "building"
	Entries uint64
	Bytes   uint64
}

// Ranked is one (key, score/distance) result from MATCH / KNN.
type Ranked struct {
	Key   []byte
	Score float64
}

// IdxCreateRange declares a plain range index:
// IDX.CREATE name ON PREFIX prefix FIELD field TYPE ty KIND range.
func (c *Client) IdxCreateRange(ctx context.Context, name, prefix, field []byte, ty IdxType) error {
	return c.IdxCreateRaw(ctx, name, bs("ON"), bs("PREFIX"), prefix, bs("FIELD"), field, bs("TYPE"), ty.tag(), bs("KIND"), bs("range"))
}

// IdxCreateRaw is the IDX.CREATE argv passthrough (everything after the
// verb) — covers MAXMEM, ANN DIM/DISTANCE/M/EF, GROUPBY, etc.
func (c *Client) IdxCreateRaw(ctx context.Context, args ...[]byte) error {
	if _, err := c.requireRemote("IDX.CREATE"); err != nil {
		return err
	}
	return c.execOK(ctx, prepend("IDX.CREATE", args))
}

// IdxDrop drops an index, returning whether it existed.
func (c *Client) IdxDrop(ctx context.Context, name []byte) (bool, error) {
	if _, err := c.requireRemote("IDX.DROP"); err != nil {
		return false, err
	}
	return c.execBool(ctx, [][]byte{bs("IDX.DROP"), name})
}

// IdxList returns the declared indexes (unknown labels skipped).
func (c *Client) IdxList(ctx context.Context) ([]IdxInfo, error) {
	if _, err := c.requireRemote("IDX.LIST"); err != nil {
		return nil, err
	}
	r, err := c.exec(ctx, argv1("IDX.LIST"))
	if err != nil {
		return nil, err
	}
	switch r.Kind {
	case KindArray:
		out := make([]IdxInfo, 0, len(r.Array))
		for _, entry := range r.Array {
			info, perr := parseIdxInfo(entry)
			if perr != nil {
				return nil, perr
			}
			out = append(out, info)
		}
		return out, nil
	case KindError:
		return nil, replyErr(r)
	default:
		return nil, unexpectedReply(r)
	}
}

// IdxQueryRange runs IDX.QUERY name RANGE min max [LIMIT n] [CURSOR c] and
// returns one page in (value, key) order. Resume with IdxPage.Cursor.
func (c *Client) IdxQueryRange(ctx context.Context, name, min, max []byte, limit int, cursor []byte) (IdxPage, error) {
	args := [][]byte{bs("RANGE"), min, max, bs("LIMIT"), itob(int64(limit))}
	if cursor != nil {
		args = append(args, bs("CURSOR"), cursor)
	}
	r, err := c.idxQueryRaw(ctx, name, args...)
	if err != nil {
		return IdxPage{}, err
	}
	return parseIdxPage(r)
}

// IdxQueryEq runs IDX.QUERY name EQ value [LIMIT n].
func (c *Client) IdxQueryEq(ctx context.Context, name, value []byte, limit int) (IdxPage, error) {
	r, err := c.idxQueryRaw(ctx, name, bs("EQ"), value, bs("LIMIT"), itob(int64(limit)))
	if err != nil {
		return IdxPage{}, err
	}
	return parseIdxPage(r)
}

// IdxQueryMatch runs IDX.QUERY name MATCH text [LIMIT n] and returns BM25
// (key, score) hits, best first.
func (c *Client) IdxQueryMatch(ctx context.Context, name, text []byte, limit int) ([]Ranked, error) {
	r, err := c.idxQueryRaw(ctx, name, bs("MATCH"), text, bs("LIMIT"), itob(int64(limit)))
	if err != nil {
		return nil, err
	}
	return parseRanked(r)
}

// IdxQueryKnn runs IDX.QUERY name KNN vector [LIMIT k], encoding vector as
// an f32 little-endian blob, and returns (key, distance) nearest first.
func (c *Client) IdxQueryKnn(ctx context.Context, name []byte, vector []float32, k int) ([]Ranked, error) {
	blob := make([]byte, len(vector)*4)
	for i, f := range vector {
		binary.LittleEndian.PutUint32(blob[i*4:], math.Float32bits(f))
	}
	r, err := c.idxQueryRaw(ctx, name, bs("KNN"), blob, bs("LIMIT"), itob(int64(k)))
	if err != nil {
		return nil, err
	}
	return parseRanked(r)
}

// IdxQueryRaw is the IDX.QUERY argv passthrough returning the raw Reply —
// the escape hatch for COMPOSE / HYBRID / GROUPS / FIELDS and any future
// query shape.
func (c *Client) IdxQueryRaw(ctx context.Context, args ...[]byte) (Reply, error) {
	if _, err := c.requireRemote("IDX.QUERY"); err != nil {
		return Reply{}, err
	}
	r, err := c.exec(ctx, prepend("IDX.QUERY", args))
	if err != nil {
		return Reply{}, err
	}
	if r.Kind == KindError {
		return Reply{}, replyErr(r)
	}
	return r, nil
}

func (c *Client) idxQueryRaw(ctx context.Context, name []byte, args ...[]byte) (Reply, error) {
	full := make([][]byte, 0, len(args)+1)
	full = append(full, name)
	full = append(full, args...)
	return c.IdxQueryRaw(ctx, full...)
}

func parseIdxPage(r Reply) (IdxPage, error) {
	if r.Kind != KindArray || len(r.Array) != 2 {
		return IdxPage{}, errProtocol("IDX.QUERY page: expected [cursor, rows]")
	}
	var page IdxPage
	switch cur := r.Array[0]; cur.Kind {
	case KindBulk:
		if string(cur.Bytes) != "0" {
			page.Cursor = cur.Bytes
		}
	default:
		return IdxPage{}, unexpectedReply(cur)
	}
	flat := r.Array[1]
	if flat.Kind != KindArray {
		return IdxPage{}, errProtocol("IDX.QUERY page: rows not an array")
	}
	if len(flat.Array)%2 != 0 {
		return IdxPage{}, errProtocol("IDX.QUERY page: odd row pair")
	}
	for i := 0; i < len(flat.Array); i += 2 {
		k, v := flat.Array[i], flat.Array[i+1]
		if k.Kind != KindBulk || v.Kind != KindBulk {
			return IdxPage{}, errProtocol("IDX.QUERY page: non-bulk row pair")
		}
		page.Rows = append(page.Rows, IdxRow{Key: k.Bytes, Value: v.Bytes})
	}
	return page, nil
}

func parseRanked(r Reply) ([]Ranked, error) {
	if r.Kind != KindArray {
		return nil, unexpectedReply(r)
	}
	out := make([]Ranked, 0, len(r.Array))
	for _, row := range r.Array {
		if row.Kind != KindArray || len(row.Array) < 2 {
			return nil, errProtocol("IDX.QUERY ranked row: expected [key, score]")
		}
		key, score := row.Array[0], row.Array[1]
		if key.Kind != KindBulk {
			return nil, unexpectedReply(key)
		}
		s, err := scoreOf(score)
		if err != nil {
			return nil, err
		}
		out = append(out, Ranked{Key: key.Bytes, Score: s})
	}
	return out, nil
}

func parseIdxInfo(entry Reply) (IdxInfo, error) {
	if entry.Kind != KindArray {
		return IdxInfo{}, unexpectedReply(entry)
	}
	var info IdxInfo
	cells := entry.Array
	for i := 0; i+1 < len(cells); i += 2 {
		label, value := cells[i], cells[i+1]
		if label.Kind != KindBulk || value.Kind != KindBulk {
			continue
		}
		switch string(label.Bytes) {
		case "name":
			info.Name = value.Bytes
		case "prefix":
			info.Prefix = value.Bytes
		case "kind":
			info.Kind = value.Str()
		case "state":
			info.State = value.Str()
		case "entries":
			n, err := parseUint(value.Bytes)
			if err != nil {
				return IdxInfo{}, err
			}
			info.Entries = n
		case "bytes":
			n, err := parseUint(value.Bytes)
			if err != nil {
				return IdxInfo{}, err
			}
			info.Bytes = n
		default:
			// forward-compatible: skip labels this version doesn't know
		}
	}
	return info, nil
}
